src/Features/Core/Portable/MakeFieldReadonly/MakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,254 @@
+                        if (IsFieldWrite(fieldReference, operationContext.ContainingSymbol))
+                        {
+                            UpdateFieldStateOnWrite(fieldReference.Field);
+                        }
+                    }, OperationKind.FieldReference);
+
+                    symbolStartContext.RegisterSymbolEndAction(symbolEndContext =>
+                    {
+                        // Report diagnostics for candidate fields that are not written outside constructor and field initializer.
+                        foreach (var kvp in fieldStateMap)
+                        {
+                            IFieldSymbol field = kvp.Key;
+                            (bool isCandidate, bool written) = kvp.Value;
+                            if (isCandidate && !written)
+                            {
+                                var option = GetCodeStyleOption(field, symbolEndContext.Options, symbolEndContext.CancellationToken);
+                                var diagnostic = DiagnosticHelper.Create(
+                                    Descriptor,
+                                    field.Locations[0],
+                                    option.Notification.Severity,
+                                    additionalLocations: null,
+                                    properties: null);
+                                symbolEndContext.ReportDiagnostic(diagnostic);
+                            }
+                        }
+                    });
+                }, SymbolKind.NamedType);
+
+                return;
+
+                // Local functions.
+                bool isCandidateField(IFieldSymbol symbol) =>
+                        symbol.DeclaredAccessibility == Accessibility.Private &&
+                        !symbol.IsReadOnly &&
+                        !symbol.IsConst &&
+                        !symbol.IsImplicitlyDeclared &&
+                        symbol.Locations.Length == 1 &&
+                        !IsMutableValueType(symbol.Type);
+
+                // Method to update the field state for a candidate field written outside constructor and field initializer.
+                void UpdateFieldStateOnWrite(IFieldSymbol field)
+                {
+                    Debug.Assert(isCandidateField(field));
+                    Debug.Assert(fieldStateMap.ContainsKey(field));
+
+                    fieldStateMap[field] = (isCandidate: true, written: true);
+                }
+
+                // Method to get or initialize the field state.
+                (bool isCandidate, bool written) TryGetOrInitializeFieldState(IFieldSymbol fieldSymbol, AnalyzerOptions options, CancellationToken cancellationToken)
+                {
+                    return fieldStateMap.GetOrAdd(fieldSymbol, valueFactory: InitializeFieldState);
+
+                    // Method to initialize the field state.
+                    (bool isCandidate, bool written) InitializeFieldState(IFieldSymbol field)
```

Comment:
1. nested nesteds is just getting really difficult to wrap my head around.
2. every time we call GetOrAdd, will this allocate a delegate around InitializeFieldState?